base/strings/string.jl

```diff
@@ -1,11 +1,125 @@
 # This file is a part of Julia. License is MIT: http://julialang.org/license
 
-## from base/boot.jl:
-#
-# immutable String <: AbstractString
-#     data::Vector{UInt8}
-# end
-#
+typealias ByteArray Union{Vector{UInt8},Vector{Int8}}
+
+## constructors and conversions ##
+
+# String constructor docstring from boot.jl, workaround for #16730
+# and the unavailability of @doc in boot.jl context.
+"""
+    String(v::Vector{UInt8})
+
+Create a new `String` from a vector `v` of bytes containing
+UTF-8 encoded characters.   This function takes "ownership" of
+the array, which means that you should not subsequently modify
+`v` (since strings are supposed to be immutable in Julia) for
+as long as the string exists.
+
+If you need to subsequently modify `v`, use `String(copy(v))` instead.
+"""
+function String(v::Array{UInt8,1})
+    ccall(:jl_array_to_string, Ref{String}, (Any,), v)
+end
+
+"""
+    unsafe_string(p::Ptr{UInt8}, [length::Integer])
+
+Copy a string from the address of a C-style (NUL-terminated) string encoded as UTF-8.
+(The pointer can be safely freed afterwards.) If `length` is specified
+(the length of the data in bytes), the string does not have to be NUL-terminated.
+
+This function is labelled "unsafe" because it will crash if `p` is not
+a valid memory address to data of the requested length.
+"""
+function unsafe_string(p::Union{Ptr{UInt8},Ptr{Int8}}, len::Integer)
+    p == C_NULL && throw(ArgumentError("cannot convert NULL to string"))
+    ccall(:jl_pchar_to_string, Ref{String}, (Ptr{UInt8}, Int), p, len)
+end
+function unsafe_string(p::Union{Ptr{UInt8},Ptr{Int8}})
+    p == C_NULL && throw(ArgumentError("cannot convert NULL to string"))
+    ccall(:jl_cstr_to_string, Ref{String}, (Ptr{UInt8},), p)
+end
+
+_string_n(n::Integer) = ccall(:jl_alloc_string, Ref{String}, (Csize_t,), n)
+
+convert(::Type{Vector{UInt8}}, s::String) = ccall(:jl_string_to_array, Ref{Vector{UInt8}}, (Any,), s)
+convert(::Type{String}, s::String) = s
+convert(::Type{String}, v::Vector{UInt8}) = String(v)
+
+## low-level functions ##
+
+pointer(s::String) = unsafe_convert(Ptr{UInt8}, s)
+pointer(s::String, i::Integer) = pointer(s)+(i-1)
+
+sizeof(s::String) = s.len
+
+"""
+    codeunit(s::String, i::Integer)
```

Comment:
If it supposed to be a general method for other encodings too, then it should be documented as `codeunit(s::AbstractString, i::Integer)`.